joomla/joomla-cms
# [5.0] Tours alias language

Pull Request for Issue # .

### Summary of Changes

1. Adds uid column to #__guidedtour table
2. Existing Joomla tours are identified as joomla-articles etc. based on the language key for the tour title.  Suggests that user adds an identifier like author/company/siteurl-tour-name - automatically creates as siteurl-tour-name when blank and checks if it is unique. 
3. Language files match this identifier naming convention - for example:  guidedtours.joomla_articles.ini and the steps in guidedtours.joomla_articles_steps.ini or extension.extension_tour_name.ini, mod_extension.module_tour_name.ini
4. Language files can be loaded from administrator/com_extension administrator/modules/mod_extension etc.
5. Parses buttons with data attributes of data-id="tour id" OR data-gt-uid="tour-uid" to allow buttons to be created that are independent of the tour id in the database

### Testing Instructions

Apply this PR
Go to System -> Manage -> Guided Tours

Test 1 - Checking identifier creation

Create a tour. Do not enter data in the identifier field.
Save the tour and check the identifier.
If you are on a local server, the identifier should look something like localhost-the-tour-name.

Test 2 - Checking identifier uniqueness

Try creating a tour with an identifier that exists for another tour.
The identifier should have a -X number attached to its name if it has been in use already.

Test 3 - Duplicating a tour. 

In the Tours view, select a tour and select 'duplicate' from the actions.
The tour language keys and step language keys should not be translated until a proper language file matches the identifier. If the tour identifier is `localhost-articles` then the language files must be `guidedtours.localhost_articles.ini` and `guidedtours.localhost_articles_steps.ini` (language files are located in administrator/language).
Note that in tours view, the language keys may be translated because identical to languages loaded from another tour.

Test 4 - Launching a tour from any location

Create a custom module in the cpanel dashboard and add this HTML to it
`
<button class="button-start-guidedtour btn btn-primary" type="button" data-gt-uid="joomla-articles"> How to create articles? </button> 
<button class="btn btn-secondary button-start-guidedtour" type="button" data-gt-uid="joomla-contacts">How to create contacts?</button>
`

Then view the tour by clicking the link in the module (which uses the identifier, not the id of the tour).

Test 5 - Making sure language files are properly used

Change the translation in the files
- administrator/language/[language used in the console]/guidedtours.joomla_articles.ini 
- administrator/language/[language used in the console]/guidedtours.joomla_articles_steps.ini
to see that these are indeed the language files that are being used.
Make sure the site is multilingual.


### Actual result BEFORE applying this Pull Request

No tour alias, no way to run a tour other than using the tour module that lists all tours.
Language strings are loaded in the dashboard for all tours, on every page where the tour module is present.

### Expected result AFTER applying this Pull Request

Tours can be triggered in other places than from the tour module.
Only languages keys that are needed are loaded.

Warning:
Third-party multilingual tours will lose their translations through this PR (the language keys are moved out of extension's language files and new language files need to follow the new naming conventions). Plan is to alert third-party developers of the change once this PR is merged so they can update their tour's language files accordingly. 

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: [New or Edit a Tour](https://docs.joomla.org/Help5.x:Guided_Tours:_New_or_Edit_Tour) - this needs updating to cover language file naming and locations
